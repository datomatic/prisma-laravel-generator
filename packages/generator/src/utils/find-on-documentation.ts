import {isEmpty} from 'lodash';
import getFromDocumentation from './get-from-documentation';

const findOnDocumentation = (
  find: string | RegExp,
  documentation?: string | undefined,
): boolean => {
  return !isEmpty(getFromDocumentation(find, documentation));
};

export default findOnDocumentation;
