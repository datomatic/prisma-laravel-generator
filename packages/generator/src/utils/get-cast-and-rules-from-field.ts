import {ConnectorType, DMMF} from '@prisma/generator-helper';
import _ from 'lodash';
import getPrismaFqcn from '../helpers/get-prisma-fqcn';
import getDatabaseFieldAttribute from './raw-schema/get-database-field-attribute';
import isFieldUuid from '../helpers/is-field-uuid';
import isFieldEnum from '../helpers/is-field-enum';
import getTableNameFromModel from '../helpers/get-table-name-from-model';
import getEnumFqcn from '../helpers/get-enum-fqcn';

const getCastAndRulesFromField = (
  field: DMMF.Field,
  model: DMMF.Model,
  enums: DMMF.DatamodelEnum[],
  rawSchema: string,
  provider?: ConnectorType,
): {
  cast: {value: string | undefined; isMethodCall: boolean};
  rules: {value: string; isMethodCall: boolean}[];
  imports: Set<string>;
} => {
  const {name: fieldName, isUnique} = field;
  const {name: modelName, uniqueFields} = model;

  const cast: {value: string | undefined; isMethodCall: boolean} = {
    value: undefined,
    isMethodCall: false,
  };
  const rules: {value: string; isMethodCall: boolean}[] = [];
  const imports = new Set<string>();

  if (isUnique) {
    imports.add('Illuminate\\Validation\\Rule');
    rules.push({
      value: `Rule::unique('${getTableNameFromModel(
        model,
      )}', '${fieldName}')->ignore($this->getKey(), $this->getKeyName())`,
      isMethodCall: true,
    });
  }
  if (!_.isEmpty(uniqueFields)) {
    const uniqueFieldsToBeAdded = _.filter(
      uniqueFields,
      fields => !_.isEmpty(fields) && fields[0] === fieldName,
    );
    if (!_.isEmpty(uniqueFieldsToBeAdded)) {
      imports.add('Illuminate\\Validation\\Rule');
      for (const ufs of uniqueFieldsToBeAdded) {
        rules.push({
          value: `Rule::unique('${getTableNameFromModel(
            model,
          )}', '${fieldName}')->ignore($this->getKey(), $this->getKeyName())->where(fn ($query) => $query->${_.chain(
            ufs,
          )
            .slice(1)
            .map(f => `where('${f}', $this->${f})`)
            .join('->')
            .value()})`,
          isMethodCall: true,
        });
      }
    }
  }

  const {type, isList, isRequired} = field;

  if (isList) {
    cast.value = 'array';
    rules.push(
      {
        value: 'array',
        isMethodCall: false,
      },
      {
        value: 'required',
        isMethodCall: false,
      },
    );
  } else {
    if (isRequired) {
      rules.push({
        value: 'required',
        isMethodCall: false,
      });
    } else {
      rules.push({
        value: 'nullable',
        isMethodCall: false,
      });
    }

    switch (type) {
      case 'String':
        cast.value = 'string';
        rules.push({
          value: 'string',
          isMethodCall: false,
        });
        break;
      case 'Boolean':
        cast.value = 'boolean';
        rules.push({
          value: 'boolean',
          isMethodCall: false,
        });
        break;
      case 'Int':
      case 'BigInt':
      case 'Bytes':
        cast.value = 'integer';
        rules.push(
          {
            value: 'numeric',
            isMethodCall: false,
          },
          {
            value: 'integer',
            isMethodCall: false,
          },
        );
        break;
      case 'Float':
        cast.value = 'double';
        rules.push({
          value: 'numeric',
          isMethodCall: false,
        });
        break;
      case 'Decimal':
        cast.value = 'string';
        rules.push({
          value: 'numeric',
          isMethodCall: false,
        });
        break;
      case 'DateTime':
        cast.value = 'immutable_datetime';
        rules.push({
          value: 'date',
          isMethodCall: false,
        });
        break;
      case 'Json':
        cast.value = 'array';
        rules.push({
          value: 'json',
          isMethodCall: false,
        });
        break;
      default: {
        const enumInfo = isFieldEnum(field, enums);
        if (enumInfo) {
          const {fqcn} = getEnumFqcn(enumInfo);
          imports.add(fqcn);
          imports.add('Illuminate\\Validation\\Rules\\Enum');

          cast.isMethodCall = true;
          cast.value = `${enumInfo.name}::class`;

          rules.push({
            value: `new Enum(${enumInfo.name}::class)`,
            isMethodCall: true,
          });
        }
        break;
      }
    }
  }

  if (isFieldUuid(field)) {
    rules.push({
      value: `uuid`,
      isMethodCall: false,
    });
  }

  const databaseMapping = getDatabaseFieldAttribute(
    fieldName,
    modelName,
    rawSchema,
  );
  if (databaseMapping) {
    switch (provider) {
      case 'mysql':
        if (
          cast.value === 'integer' &&
          [
            'UnsignedInt',
            'UnsignedSmallInt',
            'UnsignedTinyInt',
            'UnsignedBigInt',
          ].includes(databaseMapping.replace(/\(.*\)/, ''))
        ) {
          rules.push({
            value: `min:0`,
            isMethodCall: false,
          });
        }
        break;
      default:
        break;
    }
  }

  return {cast, rules, imports};
};

export default getCastAndRulesFromField;
