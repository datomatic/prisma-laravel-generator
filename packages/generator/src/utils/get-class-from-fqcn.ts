const getClassFromFQCN = (fqcn: string) => {
  const asMatch = /\s+as\s+/gi.exec(fqcn);
  if (asMatch) {
    return fqcn.slice(asMatch.index + 4).trim();
  }

  if (fqcn.includes('\\')) {
    return fqcn.slice(fqcn.lastIndexOf('\\') + 1);
  }
  return fqcn;
};

export default getClassFromFQCN;
