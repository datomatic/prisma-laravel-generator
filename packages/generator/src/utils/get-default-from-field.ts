import {ConnectorType, DMMF} from '@prisma/generator-helper';
import _ from 'lodash';
import CuidNotSupportedError from '../errors/cuid-not-supported-error';
import isFieldEnum from '../helpers/is-field-enum';
import getPrismaFqcn from '../helpers/get-prisma-fqcn';

const getDefaultFromField = (
  field: DMMF.Field,
  enums: DMMF.DatamodelEnum[],
  provider?: ConnectorType,
): {
  defaultValue?: string | boolean | number;
  isMethodCall: boolean;
  imports: Set<string>;
} => {
  let defaultValue;
  let isMethodCall = false;
  const imports = new Set<string>();

  if (typeof field.default !== 'object') {
    const enumInfo = isFieldEnum(field, enums);
    if (
      field.default &&
      enumInfo &&
      _.some(enumInfo.values, v => v.name === field.default)
    ) {
      const {fqcn} = getPrismaFqcn(enumInfo);
      imports.add(fqcn);
      defaultValue = `${enumInfo.name}::${field.default.toString()}`;
      isMethodCall = true;
    } else {
      defaultValue = field.default;
    }
  } else {
    switch (field.default.name) {
      case 'uuid':
        imports.add('Illuminate\\Support\\Str');
        defaultValue = 'Str::orderedUuid()';
        isMethodCall = true;
        break;
      case 'cuid':
        throw new CuidNotSupportedError();
      case 'now':
        if (provider === 'mongodb') {
          imports.add(`Illuminate\\Support\\Carbon`);
          defaultValue = 'Carbon::now()';
          isMethodCall = true;
        } else {
          defaultValue = undefined;
        }
        break;
      default:
        defaultValue = undefined;
        break;
    }
  }
  return {defaultValue, isMethodCall, imports};
};

export default getDefaultFromField;
