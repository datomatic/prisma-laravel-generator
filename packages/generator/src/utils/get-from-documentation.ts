import _, {isString, trim} from 'lodash';

const sanitize = (s: string) => _.chain(s).toLower().trim().value();

const getFromDocumentation = (
  find: string | RegExp,
  documentation?: string | undefined,
): string[] => {
  if (!documentation) {
    return [];
  }
  const sanitizedFind = isString(find) ? sanitize(find) : find;
  return _.chain(documentation)
    .replace(/\n/g, ',')
    .split(',')
    .map(d => trim(d))
    .filter(d =>
      isString(sanitizedFind)
        ? sanitize(d) === sanitizedFind
        : sanitize(d).match(sanitizedFind) !== null,
    )
    .value();
};

export default getFromDocumentation;
