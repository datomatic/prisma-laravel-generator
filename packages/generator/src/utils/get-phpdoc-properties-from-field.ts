import {DMMF} from '@prisma/generator-helper';
import _ from 'lodash';
import isFieldEnum from '../helpers/is-field-enum';
import isFieldReadOnly from '../helpers/is-field-read-only';
import getEnumFqcn from '../helpers/get-enum-fqcn';

// eslint-disable-next-line unicorn/prevent-abbreviations
const getPhpDocPropertiesFromField = (
  field: DMMF.Field,
  enums: DMMF.DatamodelEnum[],
  guardedFields: DMMF.Field[],
  fillableFields: DMMF.Field[],
  massAssignable: boolean,
): {phpType: string | undefined; readOnly: boolean; imports: Set<string>} => {
  const imports = new Set<string>();
  let readOnly = false;

  if (isFieldReadOnly(field, guardedFields, fillableFields, massAssignable)) {
    readOnly = true;
  }

  const {type, isList, isRequired} = field;

  let phpType;
  switch (type) {
    case 'String':
    case 'Decimal':
      phpType = 'string';
      break;
    case 'Boolean':
      phpType = 'bool';
      break;
    case 'Int':
    case 'BigInt':
    case 'Bytes':
      phpType = 'int';
      break;
    case 'Float':
      phpType = 'double';
      break;
    case 'DateTime':
      if (readOnly) {
        imports.add('Carbon\\CarbonImmutable');
        phpType = 'CarbonImmutable';
      } else {
        imports.add('Illuminate\\Support\\Carbon');
        phpType = 'Carbon';
      }
      break;
    case 'Json':
      phpType = 'array';
      break;
    default: {
      const enumInfo = isFieldEnum(field, enums);
      if (enumInfo) {
        const {fqcn} = getEnumFqcn(enumInfo);
        imports.add(fqcn);

        phpType = enumInfo.name;
      }
      break;
    }
  }

  return {
    phpType: phpType
      ? `${isRequired ? '' : '?'}${phpType}${isList ? '[]' : ''}`
      : undefined,
    readOnly,
    imports,
  };
};

export default getPhpDocPropertiesFromField;
