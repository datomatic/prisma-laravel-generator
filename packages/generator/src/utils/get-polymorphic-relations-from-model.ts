import {DMMF} from '@prisma/generator-helper';

/* TODO:
     [ ] polymorphic one-to-one
     [ ] polymorphic one-to-meny
     [ ] polymorphic many-to-many
 */

type MorphOneRelation = {
  name: string;
  className: string;
  relationshipName: string;
  typeColumn?: string; // ignore if == <relationshipName>_type
  idColumn?: string; // ignore if == <relationshipName>_id
};

type MorphToRelation = {
  name: string;
  relationshipName?: string; // ignore if == relation name (php function name)
  typeColumn?: string; // ignore if == <Str::snake(relationshipName)>_type
  idColumn?: string; // ignore if == <Str::snake(relationshipName)>_id
};

type MorphManyRelation = {
  name: string;
  className: string;
  relationshipName: string;
  typeColumn?: string; // ignore if == <relationshipName>_type
  idColumn?: string; // ignore if == <relationshipName>_id
};

type MorphToManyRelation = {
  name: string;
  className: string;
  relationshipName: string;
  pivotTable?: string; // ignore if == <any_other_word>_<Str::plural(last-word)> using words of relationshipName
  foreignPivotKey?: string; // ignore if == <relationshipName>_id
  relatedPivotKey?: string; // ignore if == <Str::snake(related->className)>_<related->keyName>
  // TODO: Error if type is not <relationshipName>_type: https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Relations/MorphToMany.php#L49
};

type MorphedByManyRelation = {
  name: string;
  className: string;
  relationshipName: string;
  pivotTable?: string; // ignore if == <any_other_word>_<Str::plural(last-word)> using words of relationshipName
  foreignPivotKey?: string; // ignore if == <Str::snake(this->className)>_<this->keyName>
  relatedPivotKey?: string; // ignore if == <relationshipName>_id
  // TODO: Error if type is not <relationshipName>_type: https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Relations/MorphToMany.php#L49
};

const getPolymorphicRelationsFromModel = (
  model: DMMF.Model,
  models: DMMF.Model[],
) => {
  const imports = new Set<string>();

  const morphOne: MorphOneRelation[] = [];
  const morphTo: MorphToRelation[] = [];
  const morphMany: MorphManyRelation[] = [];
  const morphToMany: MorphToManyRelation[] = [];
  const morphedByMany: MorphedByManyRelation[] = [];

  return {
    imports,
    morphOne,
    morphTo,
    morphMany,
    morphToMany,
    morphedByMany,
  };
};

export default getPolymorphicRelationsFromModel;
