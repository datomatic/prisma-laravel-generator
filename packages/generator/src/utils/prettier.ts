import prettier from 'prettier/standalone';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import phpPlugin from '@prettier/plugin-php/standalone';

export const prettify = (content: string): string => {
  return prettier.format(content.trim(), {
    plugins: [phpPlugin],
    parser: 'php',
  });
};

export default {prettify};
