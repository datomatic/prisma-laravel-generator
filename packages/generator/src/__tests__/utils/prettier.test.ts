import {prettify} from '../../utils/prettier';

test('prettify: with invalid php', () => {
  expect(() =>
    prettify(
      '<?php namespace App\\Enums\\Prisma;  This = is some ! invalid PHP code }',
    ),
  ).toThrowError();
});

test('prettify: valid content', () => {
  expect(
    prettify(
      '<?php namespace App\\Enums\\Prisma; enum Test { case A = "A"; case B = \'B\'; case C; }',
    ),
  ).toMatchSnapshot('prettified');
});

test('prettify: empty content', () => {
  expect(prettify('')).toMatchSnapshot('prettified');
});
