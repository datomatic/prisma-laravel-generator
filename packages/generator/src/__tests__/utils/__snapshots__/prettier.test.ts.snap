// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettify: empty content: prettified 1`] = `""`;

exports[`prettify: valid content: prettified 1`] = `
"<?php namespace App\\\\Enums\\\\Prisma;
enum Test
{
    case A = \\"A\\";
    case B = \\"B\\";
    case C;
}
"
`;
